azure-iot-device/azure/iot/device/iothub/sync_clients.py

```diff
@@ -330,7 +335,9 @@ def __init__(self, iothub_pipeline):
         :type iothub_pipeline: :class:`azure.iot.device.iothub.pipeline.IoTHubPipeline`
         """
         super(IoTHubDeviceClient, self).__init__(iothub_pipeline=iothub_pipeline)
-        self._iothub_pipeline.on_c2d_message_received = self._inbox_manager.route_c2d_message
+        self._iothub_pipeline.on_c2d_message_received = CallableWeakMethod(
+            self._inbox_manager, "route_c2d_message"
+        )
```

Comment:
Is this actually necessary here? Not that it's a problem as is, but I don't think there can be a circular dependency with inbox manager and client, because no reference to the client (or any other part of the code) is held by the inbox manager.